### PR TITLE
Use Windows User Directory for bin path

### LINKED ISSUE #199: This doesn't work on Windows despite there being a binary
I downloaded your latest version onto Windows and when I tried to use it I got the following error.

> PS ~> tfswitch --latest-stable 1.0
> [Error] : Binary path does not exist: \usr\local\bin
> [Error] : Manually create bin directory at: \usr\local\bin and try again.

I'm happy to help you figure out the Windows version of this, [nvs](https://github.com/jasongin/nvs) solves this problem in the same space.

Once this is working in Windows I'll re-open my PR to [add tfswitch to scoop](https://github.com/ScoopInstaller/Main/pull/2767) and open a PR back here for the docs.

### PR DESCRIPTION
**Pull request by MatrixCrawler**
Fixes #199

- use windows user directory for bin path as default in windows
- copy terraform.exe to bin path as symlink does not workl correctly
- update x/text package due to vulnerability
- fix various typos

### CODE REVIEW

**Comment by yermulnik** [comment]
Repo is not in active development at the moment 😢 https://github.com/warrensbox/terraform-switcher/issues/324

**Comment by MatrixCrawler** [comment]
Will have to rebase first

**Comment by MatrixCrawler** [comment]
Done with the rebasing

File: lib/defaults.go
Comment (yermulnik) [inline lib/defaults.go:1]: Would it make sense to stick with `utils.go` at once?
https://github.com/warrensbox/terraform-switcher/pull/334/files#diff-bb04fece8dd9230d7af1277a31b0b3c993917058b4343dc7f48f8bcbfe3c8463

File: lib/download_test.go [hunk 1]
```diff
@@ -21,14 +22,19 @@ func TestDownloadFromURL_FileNameMatch(t *testing.T) {
 	installPath := fmt.Sprintf(tempDir + string(os.PathSeparator) + ".terraform.versions_test")
 	macOS := "_darwin_amd64.zip"
 
-	// get current user
-	homedir, errCurr := homedir.Dir()
-	if errCurr != nil {
-		log.Fatal(errCurr)
+	home, err := homedir.Dir()
+	if err != nil {
+		log.Fatalf(`Could not detect home directory.`)
```
Comment (yermulnik) [inline lib/download_test.go:27]: Are backticks intentional (as opposite to double quotes)?

**Comment by yermulnik** [comment]
> Done with the rebasing

Still merge conflict is there

File: lib/download_test.go (hunk 1 above)
Comment (MatrixCrawler) [reply lib/download_test.go:27]: No. I will fix that.

**Review by warrensbox — APPROVED** [review]
The linux and MacOs version works.

**Comment by warrensbox** [comment]
@yermulnik can you review again?

File: lib/defaults.go
Comment (yermulnik) [reply lib/defaults.go:1]: I'm still concerned why would not use `utils.go` to sync with the other PR https://github.com/warrensbox/terraform-switcher/pull/334/files#diff-bb04fece8dd9230d7af1277a31b0b3c993917058b4343dc7f48f8bcbfe3c8463

File: lib/symlink.go [hunk 2]
```diff
@@ -1,27 +1,51 @@
...
+			os.Exit(1)
+		}
+		defer r.Close()
 
-	err := os.Symlink(cwd, dir)
-	if err != nil {
-		log.Fatalf(`
+		w, err := os.Create(dir + ".exe")
+		if err != nil {
+			log.Fatalf(`Could not create target binary: %s.`, dir+".exe")
```
Comment (yermulnik) [inline lib/symlink.go:24]: It's the same as https://github.com/warrensbox/terraform-switcher/pull/327#discussion_r1543165942 — are backticks intentional?

Comment (yermulnik) [inline lib/symlink.go:24]:
```suggestion
			log.Fatalf("Could not create target binary: %s", dir + ".exe")
```

Comment (yermulnik) [reply lib/symlink.go:24]: Let's use double quotes for consistency?

File: lib/symlink.go [hunk 3]
```diff
@@ -1,27 +1,51 @@
...
 
-//CreateSymlink : create symlink
+// CreateSymlink : create symlink or copy file to bin directory if windows
 func CreateSymlink(cwd string, dir string) {
+	// If we are on windows the symlink is not working correctly.
+	// Copy the desired terraform binary to the path environment.
+	if runtime.GOOS == "windows" {
+		r, err := os.Open(cwd)
+		if err != nil {
+			log.Fatalf(`Unable to open source binary: %s.`, cwd)
```
Comment (yermulnik) [inline lib/symlink.go:17]:
```suggestion
			log.Fatalf("Unable to open source binary: %s", cwd)
```

File: lib/symlink.go [hunk 4]
```diff
@@ -1,27 +1,51 @@
...
+		defer func() {
+			if c := w.Close(); err == nil {
+				err = c
+			}
+		}()
+		_, err = io.Copy(w, r)
+	} else {
+		err := os.Symlink(cwd, dir)
+		if err != nil {
+			log.Fatalf(`
```
Comment (yermulnik) [inline lib/symlink.go:36]: Can backticks be converted to double quotes for consistency (given there are double quotes inside the message — can they be escaped or single quotes used)?

File: lib/symlink_test.go [hunk 5]
```diff
@@ -1,51 +1,74 @@
...
+	testSymlinkDest := "/test-tfswitcher-dest"
 	testSymlinkSrc := "/test-tfswitcher-src"
+	if runtime.GOOS == "windows" {
+		testSymlinkSrc = "/test-tfswitcher-src.exe"
+	}
 
-	testSymlinkDest := "/test-tfswitcher-dest"
+	home, err := homedir.Dir()
+	if err != nil {
+		log.Fatalf(`Could not detect home directory.`)
```
Comment (yermulnik) [inline lib/symlink_test.go:26]:
```suggestion
		log.Fatalf("Could not detect home directory.")
```

File: lib/symlink_test.go [hunk 6]
```diff
@@ -1,51 +1,74 @@
...
+	symlinkPathSrc := filepath.Join(home, testSymlinkSrc)
+	symlinkPathDest := filepath.Join(home, testSymlinkDest)
 
-	homedir, errCurr := homedir.Dir()
-	if errCurr != nil {
-		log.Fatal(errCurr)
+	// Create file for test as windows does not like no source
+	create, err := os.Create(symlinkPathDest)
+	if err != nil {
+		log.Fatalf(`Could not create test dest file for symlink at %v`, symlinkPathDest)
```
Comment (yermulnik) [inline lib/symlink_test.go:34]:
```suggestion
		log.Fatalf("Could not create test dest file for symlink at %v", symlinkPathDest)
```

**Comment by yermulnik** [comment]
> @yermulnik can you review again?

It looks good to me apart 1) `defaults.go` vs `utils.go` and 2) backticks vs double quotes nitpicks.

ps: Thanks for pinging me as I was waiting for @MatrixCrawler to re-request review once he's done with updates. Apologies for delay.

**Comment by yermulnik** [comment]
> > @yermulnik can you review again?
> 
> It looks good to me apart 1) `defaults.go` vs `utils.go` and 2) backticks vs double quotes nitpicks.

@MatrixCrawler @warrensbox Please let me know whether this is meant to be worked on in a new PR and I'll be happy to add my approval to get this merged and released.

**Comment by MatrixCrawler** [comment]
> > > @yermulnik can you review again?
> > 
> > 
> > It looks good to me apart 1) `defaults.go` vs `utils.go` and 2) backticks vs double quotes nitpicks.
> 
> @MatrixCrawler @warrensbox Please let me know whether this is meant to be worked on in a new PR and I'll be happy to add my approval to get this merged and released.

I would see to this in a later PR as i introduce the defaults and the utils in two new PR's. There should be a consolidation, refactoring at a later point.
Just my 2 cents

**Review by yermulnik — APPROVED** [review]

**Comment by yermulnik** [comment]
> I would see to this in a later PR as i introduce the defaults and the utils in two new PR's. There should be a consolidation, refactoring at a later point.

I don't see why the consolidation cannot be done in this PR given both are authored by you and this could reduce a number of extra PRs just to merge two files with helper functions 🤷🏻 
Either way I approved this PR, @MatrixCrawler please go ahead with merging it whenever you're up for it 🤝

**Review by warrensbox — APPROVED** [review]

**Pull request merged**